cli/updater/updater.go

```diff
@@ -0,0 +1,138 @@
+)
+
+var tr = i18n.Tr
+
+// CheckForUpdate returns the latest available version if greater than
+// the one running and it makes sense to check for an update, nil in all other cases
+func CheckForUpdate(currentVersion *semver.Version) *semver.Version {
+	if !shouldCheckForUpdate(currentVersion) {
+		return nil
+	}
+
+	return checkForUpdate(currentVersion)
+}
+
+// ForceCheckForUpdate always returns the latest available version if greater than
+// the one running, nil in all other cases
+func ForceCheckForUpdate(currentVersion *semver.Version) *semver.Version {
+	return checkForUpdate(currentVersion)
+}
+
+func checkForUpdate(currentVersion *semver.Version) *semver.Version {
+	defer func() {
+		// Always save the last time we checked for updates at the end
+		inventory.Store.Set("updater.last_check_time", time.Now())
+		inventory.WriteStore()
+	}()
+
+	latestVersion, err := semver.Parse(getLatestRelease())
+	if err != nil {
+		return nil
+	}
+
+	if currentVersion.GreaterThanOrEqual(latestVersion) {
+		// Current version is already good enough
+		return nil
+	}
+
+	return latestVersion
+}
+
+// NotifyNewVersionIsAvailable prints information about the new latestVersion
+func NotifyNewVersionIsAvailable(latestVersion string) {
+	feedback.Errorf("\n\n%s %s → %s\n%s",
+		color.YellowString(tr("A new release of Arduino CLI is available:")),
+		color.CyanString(globals.VersionInfo.VersionString),
+		color.CyanString(latestVersion),
+		color.YellowString("https://arduino.github.io/arduino-cli/latest/installation/#latest-packages"))
+}
+
+// shouldCheckForUpdate return true if it actually makes sense to check for new updates,
+// false in all other cases.
+func shouldCheckForUpdate(currentVersion *semver.Version) bool {
+	if strings.Contains(currentVersion.String(), "git-snapshot") || strings.Contains(currentVersion.String(), "nightly") {
+		// This is a dev build, no need to check for updates
+		return false
+	}
+
+	if !configuration.Settings.GetBool("updater.enable_notification") {
+		// Don't check if the user disabled the notification
+		return false
+	}
+
+	if inventory.Store.IsSet("updater.last_check_time") && time.Since(inventory.Store.GetTime("updater.last_check_time")).Hours() < 24 {
+		// Checked less than 24 hours ago, let's wait
+		return false
+	}
+
+	// Don't check when running on CI or on non interactive consoles
+	return !isCI() && configuration.IsInteractive && configuration.HasConsole
+}
+
+// based on https://github.com/watson/ci-info/blob/HEAD/index.js
+func isCI() bool {
+	return os.Getenv("CI") != "" || // GitHub Actions, Travis CI, CircleCI, Cirrus CI, GitLab CI, AppVeyor, CodeShip, dsari
+		os.Getenv("BUILD_NUMBER") != "" || // Jenkins, TeamCity
+		os.Getenv("RUN_ID") != "" // TaskCluster, dsari
+}
+
+// getLatestRelease queries the official Arduino download server for the latest release,
+// if there are no errors or issues a version string is returned, in all other case an empty string.
+func getLatestRelease() string {
```

Comment:
We can make this function return a `*semver.Version` (or `nil` if fails or not valid semver)

Reply:
I don't think that would help that much, this `getLatestRelease` is only used in `checkForUpdate` and there we handle eventual errors in parsing the version string, if I change `getLatestRelease` to return a `*semver.Version` I'd have to check for the parsing error inside it and handle the `nil` version in `checkForUpdate` too. 

Don't seem much of an improvement to me really. 🤔